### DISCUSSION TITLE
Is LVGL usable with I2C Displays?

### DISCUSSION THREAD

**Discussion by markusdd** [arendst/Tasmota#20351] [Q&A] [answered]
While I wait for my larger displays to arrive from PRC I was trying to get warm with berry and LVGL using one of my devkits and a SSD1306 OLED display which uses i2c.

I can 'dump' my lvgl code in the console and it is accepted, but nothing shows on the display.

Is thsi a supported mode or do only 'fat' SPI displays and RGB mode displays work?

**Comment by s-hadinger** [comment, marked as answer]
I don't see any reason why it shouldn't work with I2C display if you are using the universal driver and the correct `display.ini`. Anyways you should see a splash screen at start with the Tasmota logo. If you don't see it then the driver is not well configured for your display

**Comment by markusdd** [comment]
That was the missing link. So LVGL requires the universal driver, I was trying to use the included SSD1306 driver.
So I just have to search for a template that uses this display, because I don't know the sequence for the driver but these displays are very common

**Reply by sfromis** [reply]
Not in the template, but there are already display descriptor files for SSD1306, select one to match the dimensions.
https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc

Still, to enable the universal display driver, you do need `Option A 3` configured on an unused gpio.

**Reply by sfromis** [reply]
FTR, configured a 128x64 SSD1306 with universal display and it worked to briefly display the display name (SSD1306) at boot, but not the Tasmota logo, likely due to it not fitting in the small display. And with a simple HASPmota setup to display clock and date, this worked too; thus also LVGL.

**Reply by markusdd** [reply]
currently trying that.

I have now flashed the official tasmota32-lvgl binary but with it the system is strangely almost unusable. It is a standard ESP32 devkit, and it ran fine with my custom configured build but somehow this one slows it to an absolute crawl.

**Reply by markusdd** [reply]
Really odd. I have put the display.ini file into the filesystem, but somehow it is not picked up. (yes I rebooted) 
![grafik](https://github.com/arendst/Tasmota/assets/25175069/19171fbe-332c-4682-960f-97cce32da7e7)

**Reply by s-hadinger** [reply]
Did you set a gpio with `OPTION A3`?

This is required to enable universal driver

**Reply by markusdd** [reply]
![097375c2e441bcc0-600x338](https://github.com/arendst/Tasmota/assets/25175069/156843a0-5e3f-4da9-92d0-b4383d2681f5)

Of course I did not haha

**Reply by sfromis** [reply]
BTW. Extending my HASPmota test case with an image worked too, however with a bit of flickering. And suddenly the screen flipped top/bottom. Maybe the uDisplay SSD1306 support has not been tested much.

**Reply by markusdd** [reply]
Is there actually an API reference for the Berry LVGL bindings? And can these be tested offline on Desktop like LVGL itself?

I have the C-Code (part manual, part from SquareLineStudio export) and I am looking into how to convert it. A lot of stuff is straightforward but for some things I am missing the correct function signatures.
I have found the berry manual and so on but I am lacking just a plain old AI reference to look up such stuff.

**Reply by sfromis** [reply]
You can mostly base it on the LVGL reference for C, and:
https://tasmota.github.io/docs/LVGL/#converting-c-lvgl-to-berry
https://tasmota.github.io/docs/LVGL/#whats-implemented-and-whats-not

**Reply by s-hadinger** [reply]
There isn't per se. Considering the vast number of signatures, it would need to be automated.

Signatures are here: https://github.com/arendst/Tasmota/blob/development/lib/libesp32_lvgl/lv_binding_berry/generate/be_lv_c_mapping.h

They follow the conventions of Berry_mapping: https://github.com/berry-lang/berry_mapping

Constants and top-level module functions are here: https://github.com/arendst/Tasmota/blob/development/lib/libesp32_lvgl/lv_binding_berry/generate/be_lvgl_module.c

It's mostly straightforward from C, with some additional work for some specific structures. You can see advanced examples here (ported from LVGL doc): https://github.com/arendst/Tasmota/tree/development/tasmota/berry/lvgl_examples/lvgl_examples_from_doc

**Reply by markusdd** [reply]
I see, that helps for the most part.

btw: Is there any way to also integrate C-code for LVGL? If you have the C-source from the generator already it seems kind of silly to translate it again.

**Comment by markusdd** [comment]
@sfromis as you seem to have a testcase with this display type as well: I have a strange issue right now.
Setting the background works fine

```py
lv.start()
hres = lv.get_hor_res()      
vres = lv.get_ver_res()      


scr = lv.scr_act()            # default screen object
f20 = lv.montserrat_font(20)  # load embedded Montserrat 20
scr.set_style_bg_color(lv.color(0xFFFFFF), lv.PART_MAIN | lv.STATE_DEFAULT)
line_basement = lv.line(scr);
top_y = 4
bottom_y = 60
left_x = 8
right_x = 120
line_basement_points = []
line_basement_points.push(lv.point(8, 4))
line_basement_points.push(lv.point(120, 4))
line_basement_points.push(lv.point(120, 60))
line_basement_points.push(lv.point(8, 60))
line_basement.set_points(lv.point_arr(line_basement_points))
# style the line
style_line = lv.style()
style_line.init()
style_line.line_width(3)
style_line.line_color(lv.color(0x000000))
style_line.line_rounded(true)
line_basement.add_style(style_line, 0)
```

but when I past the whole code that shoudl draw the line the device reboots or does nothing. very odd.

Any ideas? I need the line feature to work but I seem to be missing something.

**Comment by s-hadinger** [comment]
Actually you can't do `lv.point(8, 4)` because of internal implementation.

You need to do:
```
p0 = lv.point()
p0.x = 8
p0.y = 4
```

I suggest you use a wrapper function to make it easier.

**Reply by markusdd** [reply]
oh ok that is kind of verbose. (and probably needs an example and doc update because that is kind of unexpected)

Alos, why is this not producing any kind of error?

**Reply by s-hadinger** [reply]
Becaus `lv.point` is a mapping of a C struct, not a real object. In such case, the first parameter if present is a pointer to a memory location containing a pre-existing value. Here you passes `8` as a pointer to location 0x00008 which creates a crash when dereferenced.

```
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x401407c7  PS      : 0x00060b30  A0      : 0x80145cb9  A1      : 0x3ffd37a0
A2      : 0x3f810acb  A3      : 0x3f810ad6  A4      : 0x00000008  A5      : 0x0000000c
A6      : 0x00000004  A7      : 0x00000001  A8      : 0x3f810acb  A9      : 0x3ffd3790
A10     : 0x3f493753  A11     : 0x3f4603a9  A12     : 0x00000008  A13     : 0x3f810acc
A14     : 0x74796200  A15     : 0x00272873  SAR     : 0x00000008  EXCCAUSE: 0x0000001c
EXCVADDR: 0x00000008  LBEG    : 0x4008c269  LEND    : 0x4008c28b  LCOUNT  : 0xffffffff
```

See: `EXCVADDR: 0x00000008`

If you look at the `C` code, it's `lv_point_t p = {8, 4};`, behaving the same way.

**Reply by markusdd** [reply]
yes, but I was hoping something more concise like this is possible (from my C code):
```c
static lv_point_t line_basement_points[] = {
        {left_x, top_y},               // top left origin
        {left_x + 450, top_y},         // north wall
        {left_x + 450, 190},           // outer door wall
        ...
    };
    lv_line_set_points(line_basement, line_basement_points, 30);


```

**Reply by sfromis** [reply]
When it comes to the verbosity, you could have a tiny helper function like:
```
def lvpoint(x,y)
  point = lv.point()
  point.x = x
  point.y = y
return point
...
var line_basement_points = lv.point_arr([
  self.lvpoint(8, 4),
  self.lvpoint(120, 4),
  self.lvpoint(120, 60),
  self.lvpoint(8, 60)])
```

Above assumes a class instance context; if not, that `self.` is not relevant.

**Reply by markusdd** [reply]
true. yet not needed in C.

would it make sense to somehow add this to the bindings?
Berry is supposed to be 'easier' to use and more concise than C, so it would really elevante the experience if this would work

my_line.set_points([[1.1],[2,4],[1,4],[1,1])

I have a pretty 'braindead' python script now which does most of the lifting to get me some close-to-berry code out of my C code base that already works in the PC simulator, but I am experiencing that the devil is really in the Details here.

And picking the berry signatures out of a generated C-file is also not exactly comfortable, so I would guess anyone attempting something bigger here will have quite a hard time figuring this all out.

Many people would probably use Squareline studio to get their initial UI prototype and the road is somewhat rocky right now to put already working C-Code into yet another language.

I am pondering if having the option to add-in the UI code in C as an extension and then just doing the callback handling etc in berry would actually be more user friendly, because the UI code in berry is adding hardly any extra value.

Other option would be to leverage the header-extraction scripting and providing a more sophisticated 'transpiler' that takes the C-UI-Code and except for more complicated coinstruct turns it into berry code. Because fixing all of this manually with search/replace for 1000 lines of code is no fun haha xD

**Reply by sfromis** [reply]
Of course, that `set_points` would be trivial to create as a small helper function.

Anything with C++ code would be a *major* slowdown to the quick process of repeatedly adjusting things without needing cumbersome steps to build Tasmota and flash a new binary.

Anyway, my preference with display stuff in Tasmota would be to just not have the laborious handling of details with LVGL, and instead use HASPmota. While this does not have the full flexibility of LVGL, it works great for typical simple display setups.

**Reply by markusdd** [reply]
the latter is true compared to 'dumping' a berry script, but right now (at least I have not seen it) there is no way of offline prototyping the GUI with berry.

For my project I have used this as a template: https://github.com/lvgl/lv_port_pc_vscode
and with that you can completely build the UI on PC until it looks and works as desired and only then you need to go onto the actual device. But that is only fast if either we can make berry work with that PC simulation or have an automated translation.

Because if somethign is wrong debugging berry on the device is not exactly easy I have to say, with the PC sim you have basically the full debugger at your hands, which is really nice.

And I have looked into HASPmota and the OpenHASP idea but that suffers exactly from the same issue: all the nice LVGL ecosystem like Squareline Studio etc does not work. And I have to say that writing my UI in JSON is not exactly what I find intuitive, I think LVGL is actually a very descriptive framework, while still being very flexible. Not sure that this abstraction is really what I would prefer.

But that is the nice thing: you can actually have both.

Just the C-to-berry (or HASPmota for that matter) thing is what hurts right now. Because it creates double work with no extra value. So to really make this take off, which I would like to see, we probably need to somehow tackle this transpiler issue.

**Reply by Jason2866** [reply]
Sorry for the dumb question, what is the benefit of converting working C LVGL code?

**Reply by markusdd** [reply]
that kind of is the point of the discussion^^

pro: you get to rapidly prototype further as you can just dump the berry code into tasmota console instead of recompile
also pro: you can then leverage the whole tasmota infrastructure in the code for callbacks, wifi, filesystem etc in the berry code and don't need to code this all yourself, this is actually my main motivation

contra: as you say, it is converting working UI code into yet another language

my gut feeling is actually that the ability to compile just the UI part, hand it to tasmota as a 'plug-in' and then do the rest in berry would be super nice.

**Reply by Jason2866** [reply]
You are nominated to write the needed transpiler ;-)

**Reply by markusdd** [reply]
![I-have-three-dollars-patrick-meme-1](https://github.com/arendst/Tasmota/assets/25175069/d3b1a274-b3e8-4fc7-80a1-684c210919a9)


```py
with open("ui_Screen1.c", "r") as f:
    with open("ui.be", "w") as tf:
        code = f.readlines()
        for l in code:
            if l.startswith("\n"):
                tf.write(l)
                continue
            if "#include" in l:
                continue
            if l.startswith("{"):
                continue
            if l.startswith("//"):
                continue
            if l.startswith("}"):
                tf.write("end\n")
                continue
            if l.startswith("void") and l.endswith("(void)\n"):
                tf.write("def ui_init()\n")
                continue
            if l.startswith("    "):
                line = (
                    l.replace(";\n", "\n")
                    .replace("; ", " ")
                    .replace("///", "#")
                    .replace("lv_", "lv.")
                    .replace("LV_", "lv.")
                    .replace("color_hex", "color")
                    .lstrip(" ")
                )
                tf.write(line)

```

**Reply by barbudor** [reply]
Python to convert C into Berry 🤣
👍

**Reply by markusdd** [reply]
By the way: How to integrate more Font sizes of Montserrat?

For my final project for the 7 inch screen I need size 40 but it is missing.

**Reply by sfromis** [reply]
In general, you can load scalable fonts into the file system, and use them. I'm using CenturyGothic.ttf when I wanted a larger size, with the size appended to the font file name.

**Reply by markusdd** [reply]
Is there docs on how to do this? And then how to use them in berry.

But frankly, I am fine with Montserrat and it is all right there in LVGL, I just need to know what to enable to it gets compiled.

EDIT: tasmota\lvgl_berry\tasmota_lv_conf.h this seems to be the file. Can I override those defines in my user config directly or do I need to hack it in my clone directly in this file?

**Reply by sfromis** [reply]
Using file system fonts is simply what I already mentioned. Storing a TrueType font in the file system, and using the file name along with a hyphen and size. `"text_font":"CenturyGothic.ttf-31"`
https://tasmota.github.io/docs/HASPmota/#attributes-related-to-text-content
Seems to work without that suggested `A:` before the file name.

AFAICT, there is some config scheme causing `#define` of names like `CONFIG_LV_FONT_MONTSERRAT_40` to  select embedded fonts, but probably with some sort of dynamic building of the name, which I did not spot. `lv_conf_internal.h`

**Reply by markusdd** [reply]
Ok but what you describe is for haspmota, how to use them in berry?

Meanwhile I sort of helped myself by hacking the lv_conf in the tasmota repo, enabling the 40-size and recompiling the firmware.bin, that works as well.

still using arbitrary ttfs in berry would also be of interest.

**Reply by sfromis** [reply]
HASPmota **is** Berry too, and is based on using LVGL. Do read the docs, like:
https://tasmota.github.io/docs/LVGL/#freetype-fonts-support

**Reply by markusdd** [reply]
Ah. yet another page haha. But that helps.

Currently thinking on how we could consolidate that info a bit.

While we are at it: is there a refernce for all the commands that hide behind the 'tasmota.' object in berry?

This whole endeavor suffers a little from the fact that there usually is some docs but it is spread across many locations.

**Reply by sfromis** [reply]
Obviously: https://tasmota.github.io/docs/Berry/#tasmota-object

And the docs site does have a search at the top of all pages. Usually works reasonably for single-word searches, for multiple words better prefix each word with a `+`

**Reply by markusdd** [reply]
perfect!

it would be lovely to somehow cross-reference the lvgl and tasmota extensions to berry in the main berry docs, that would make finding these a little easier I think.

Once I have all of this working I will think about how I can contribute a bit to the docs so stuff I fell across can be improved.

Thanks for all the patience and help here :D

**Reply by markusdd** [reply]
Hey guys.

To give back a little for all the help I published my still dumb but less dumb converter script for Squareline Studio UI files.
For my example it produces 100% valid berry code, for more complex scanrios mileage may vary but it should get you a lot of the way there without having to resort to changing hunderds of lines yourself using serach/replace.

Hope you enjoy it and it is of use to the community. Feel free to link to it if you want.

https://github.com/markusdd/squareline_lvgl_to_berry/

**Reply by sfromis** [reply]
I posted the link on Discord too
